### PR TITLE
Fix undefined array key issue in class-wc-tracks.php for role handling

### LINKED ISSUE #53848: undefined array key in class-wc-tracks.php after update to 9.5.1
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

an error message is displayed in the backend. the error message. The error message is displayed in various places in the backend.
The message is always the same: Undefined array key 0 in /wp-content/plugins/woocommerce/includes/tracks/class-wc-tracks.php on line 78
the easiest way to show the message is to show "designs" --> "themes" in backend.
See here: https://drive.google.com/file/d/1OppQdcbS2EUs4qwrcitNXJ1QVGzqWFql/view?usp=drive_link

it seems a added function get_role_details” in class-wp-tracks.php is the problem. See:
https://drive.google.com/file/d/17Agy6OAhUE6UYJcFfTos5BT0q3e1HAbY/view?usp=drive_link

In previos versions there is no such function. At the moment I cannot see any problems with the shop's operations. (no message on website - only in backend) 

### Expected behavior

no displayed message

### Actual behavior

some backend pages shows the error message "Undefined array key 0 in /wp-content/plugins/woocommerce/includes/tracks/class-wc-tracks.php on line 78"

### Steps to reproduce

1.) Go to backend
2.) click Design 

Error was shown on top of the page

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.5.1
Legacy REST API Package Version: Auf dieser Website ist das alte REST-API-Plugin nicht installiert.
Action Scheduler Version: ✔ 3.8.2
Log Directory Writable: ✔
WP Version: 6.7.1
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: de_DE
External object cache: –

### Server Environment ###

Server Info: LiteSpeed
PHP Version: 8.1.30
PHP Post Max Size: 64 MB
PHP Time Limit: 6000
PHP Max Input Vars: 1000
cURL Version: 7.74.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 10.11.6-MariaDB-0+deb12u1
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

attachment: 9202
avia_framework_post: 17
custom_css: 2
customize_changeset: 2
document_template: 27
forum: 2
frm_form_actions: 5
frm_styles: 1
invoice: 1031
nav_menu_item: 372
page: 114
portfolio: 17
post: 64
product: 71
product_variation: 22
revision: 172
shop_order: 4129
shop_order_refund: 84
topic: 2
vc_grid_item: 1
vc4_templates: 3
was: 73
woo_discount: 1
woo_discount_cart: 1
wp_global_styles: 1
wp_navigation: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (40) ###

Akismet Anti-spam: Spam Protection: von Automattic - Anti-Spam-Team – 5.3.5
Classic Editor: von WordPress-Mitwirkende – 1.6.7
Code Snippets: von Code Snippets Pro – 3.6.6.1
Complianz | GDPR/CCPA Cookie Consent: von Really Simple Plugins – 7.1.5
CryptX: von Ralf Weber – 3.4.5.3
Display Posts: von Bill Erickson – 3.1.1
Envato Market: von Envato – 2.0.12
Flexible Checkout Fields: von WP Desk – 4.1.15
Formidable Forms Pro: von Strategie11 – 6.16.4
Formidable Views: von Strategy11 – 4.0.05
Formidable Forms: von Strategy11 Form Builder Team – 6.16.3
XML Sitemap Generator for Google: von Auctollo – 4.1.21
Include Me: von Stefano Lissa – 1.3.2
Insert Pages: von Paul Ryan – 3.9.1
WPBakery Page Builder: von Michael M - WPBakery.com – 8.1
LiteSpeed Cache: von LiteSpeed Technologies – 6.5.4
Menu Icons: von ThemeIsle – 0.13.16
Mollie-Zahlungen für WooCommerce: von Mollie – 7.9.0
One Stop Shop für WooCommerce: von vendidero – 1.6.3
Slider Revolution: von ThemePunch – 6.7.23
Shariff Wrapper: von Jan-Peter Lambeck & 3UU – 4.6.14
WPML Multilingual CMS: von OnTheGoSystems – 4.6.15
Sucuri Security - Auditing, Malware Scanner and Hardening: von Sucuri Inc. – 1.9.7
Themeone Shortcodes: von ThemeOne – 3.7
Discount Rules Core: von Flycart – 2.6.6
WooCommerce.com Update Manager: von Automattic – 1.0.3
WooCommerce Advanced Shipping: von Jeroen Sormani – 1.1.4
WooCommerce Composite Products: von Woo – 10.2.2
Germanized für WooCommerce Pro: von vendidero – 4.1.5
Germanized für WooCommerce: von vendidero – 3.18.5
WooCommerce Multilingual & Multicurrency: von OnTheGoSystems – 5.3.9
WooCommerce: von Automattic – 9.5.1
WordPress Post Tabs: von Internet Techies
Tejaswini Deshpande – 1.6.3

Mail Bank: von Tech Banker – 4.0.14
Connect Matomo: von André Bräkling – 1.0.30
Shield Security: von Shield Security – 20.1.2
WPML CMS Nav: von OnTheGoSystems – 1.5.5
WPML Media Translation: von OnTheGoSystems – 2.7.5
WPML Sticky Links: von OnTheGoSystems – 1.5.4
WPML String Translation: von OnTheGoSystems – 3.2.17

### Inactive Plugins (3) ###

vendidero Helper: von vendidero – 2.2.1
WP File Manager: von mndpsingh287 – 8.0
Yoast Duplicate Post: von Enrico Battocchi & Team Yoast – 4.5

### Dropin Plugins () ###

advanced-cache.php: advanced-cache.php

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: composite (composite)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 Tage
Level threshold: –
Log directory size: 961 kB

### WC Pages ###

Shop-Basis: #6551 - /shop-content/
Warenkorb: #26504 - /warenkorb-2/ -  Enthält den [woocommerce_cart]-Shortcode
Kasse: #27491 - /kasse-3/ -  Enthält den [woocommerce_checkout]-Shortcode
Mein Konto: #27492 - /mein-konto-3/
Allgemeine Geschäftsbedingungen: #3683 - /agb/

### Theme ###

Name: Awekas Child Theme
Version: 1.0
Author URL: 
Child Theme: ✔
Parent Theme Name: Enfold
Parent Theme Version: 6.0.8
Parent Theme Author URL: https://kriesi.at
Theme type: Classic-Theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Composite Products ###

Database Version: 10.2.2
Loopback Test: ✔
Template Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: product-data-views
experimental-blocks
coming-soon-newsletter-template
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings
use-wp-horizon
add-to-cart-with-options-stepper-layout

Daily Cron: ✔ Next scheduled: 2024-12-19 11:48:33 +01:00
Options: ✔
Notes: 6
Onboarding: completed

### Action Scheduler ###

Abgebrochen: 1
Oldest: 2024-11-21 06:49:23 +0100
Newest: 2024-11-21 06:49:23 +0100

Abgeschlossen: 1.104
Oldest: 2024-11-17 19:40:04 +0100
Newest: 2024-12-18 17:51:04 +0100

Fehlgeschlagen: 6
Oldest: 2023-06-23 09:13:42 +0200
Newest: 2024-12-18 17:52:05 +0100

Ausstehend: 8
Oldest: 2024-12-18 22:44:06 +0100
Newest: 2024-12-22 09:58:07 +0100


### Status report information ###

Generated at: 2024-12-18 18:34:47 +01:00
```
</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/53848

This PR addresses an issue where an undefined array key error occurs in class-wc-tracks.php when handling user roles. The error arises due to the assumption that roles are a zero-indexed array, but in some cases, they may be an associative array. This update makes the role handling code more robust to avoid such errors.

I wasn't able to reproduce the issue but I added a test case to cover the scenario where the roles are not sequential to confirm the fix.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a site with this branch
2. Install `Code Snippets` plugin
3. Add the following code snippet:

```php
add_action( 'init', function() {
    if ( class_exists( 'WC_Tracks' ) && method_exists( 'WC_Tracks', 'get_role_details' ) ) {
        $user = wp_get_current_user();
        $user->roles = array(
            2 => 'administrator',
        );
        $role_details = WC_Tracks::get_role_details( $user );
		error_log( "role_details: " . print_r( $role_details, true ));
    }
});
```

4. Go to `wp-admin/themes.php` and confirm no error is thrown.
5. Check the `debug.log` file and confirm role details are correctly retrieved.

```bash
role_details: Array
(
    [role] => administrator
    [can_install_plugins] => 1
    [can_activate_plugins] => 1
    [can_manage_woocommerce] => 1
)
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/tracks/class-wc-tracks.php [hunk 1]
```diff
@@ -75,7 +75,7 @@ public static function get_server_details() {
 	 */
 	public static function get_role_details( $user ) {
 		return array(
-			'role'                   => ! empty( $user->roles ) ? $user->roles[0] : '',
+			'role'                   => ! empty( $user->roles ) ? reset( $user->roles ) : '',
```
Comment (rjchow) [inline plugins/woocommerce/includes/tracks/class-wc-tracks.php:78]: interesting usage of `reset()`!

**Review by rjchow — APPROVED** [review]
lgtm, thanks for fixing this quickly! does it need a CFE?

File: plugins/woocommerce/includes/tracks/class-wc-tracks.php (hunk 1 above)
Comment (adrianduffell) [reply plugins/woocommerce/includes/tracks/class-wc-tracks.php:78]: Could maybe also use `array_key_first` for readability 🤔

Comment (chihsuan) [reply plugins/woocommerce/includes/tracks/class-wc-tracks.php:78]: Thanks for the suggestion! The `array_key_first` retrieves an array's first **key**, not the **first** value. So, it wouldn't work in this case.

**Review by ilyasfoo — APPROVED** [review]
Nice fix! Tests well, LGTM! 🚢

**Pull request merged**